### DISCUSSION TITLE
MPCD using MPI

### DISCUSSION THREAD

**Discussion by sgoh446** [glotzerlab/hoomd-blue#1887] [Support]
I am trying to run the hoomd-mpcd routine using multiple CPUs or GPUs.
I managed to run mpcd simulations with a single rank, both on CPU and on GPU, 
but the script I used does not run on multiple ranks.
As far as I understand, I can simply run the same script written for a single core, can't I?
I would be very grateful if you could help me.

Here is the code (test_hoomd_mpcd.py) that I am using:
```
import hoomd
import numpy

device = hoomd.device.GPU()
simulation = hoomd.Simulation(device=device, seed=1)

L=64
snapshot = hoomd.Snapshot()
snapshot.configuration.box = [L,L,L,0,0,0]

density = 10
rng = numpy.random.default_rng(seed=42)

snapshot.mpcd.types = ['A']
snapshot.mpcd.N = int(density * L * L * L)
snapshot.mpcd.position[:] = [0.5*L, 0.5*L, 0.5*L] * rng.uniform(low=-1,high=1,size=(snapshot.mpcd.N,3))

kT = 1
velocity = rng.normal(0.0, numpy.sqrt(kT), (snapshot.mpcd.N, 3))
velocity -= numpy.mean(velocity, axis=0)
snapshot.mpcd.velocity[:] = velocity

simulation.create_state_from_snapshot(snapshot)

integrator = hoomd.mpcd.Integrator(dt=0.02)
integrator.collision_method = hoomd.mpcd.collide.StochasticRotationDynamics(
    period=1, angle=130, kT=kT
)

integrator.streaming_method = hoomd.mpcd.stream.Bulk(
    period=integrator.collision_method.period
)

integrator.mpcd_particle_sorter = hoomd.mpcd.tune.ParticleSorter(trigger=2)
simulation.operations.integrator = integrator

simulation.run(2)
device.notice(f'{simulation.tps}')
```

On the HPC I am using, I run
`srun --exact -n 4 --gres=gpu:4 python test_hoomd_mpcd.py`

And the error messages I got:
```
Traceback (most recent call last):
  File "/p/project1/cdim/hoomd/test_hoomd_mpcd.py", line 14, in <module>
Traceback (most recent call last):
  File "/p/project1/cdim/hoomd/test_hoomd_mpcd.py", line 14, in <module>
Traceback (most recent call last):
  File "/p/project1/cdim/hoomd/test_hoomd_mpcd.py", line 14, in <module>
    snapshot.mpcd.types = ['A']
    snapshot.mpcd.types = ['A']
    ^^^^^^^^^^^^^
  File "/p/project1/cdim/hoomd/hoomd/snapshot.py", line 415, in mpcd
    ^^^^^^^^^^^^^
  File "/p/project1/cdim/hoomd/hoomd/snapshot.py", line 415, in mpcd
    snapshot.mpcd.types = ['A']
    ^^^^^^^^^^^^^
  File "/p/project1/cdim/hoomd/hoomd/snapshot.py", line 415, in mpcd
    raise RuntimeError("Snapshot data is only present on rank 0")
RuntimeError: Snapshot data is only present on rank 0
    raise RuntimeError("Snapshot data is only present on rank 0")
RuntimeError: Snapshot data is only present on rank 0
    raise RuntimeError("Snapshot data is only present on rank 0")
RuntimeError: Snapshot data is only present on rank 0
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 2 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 3 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

Thank you very much!!!

**Comment by joaander** [comment]
As the error message states, the snapshot data is present only on rank 0. The [MPI tutorial](https://hoomd-blue.readthedocs.io/en/v4.8.2/tutorial/03-Parallel-Simulations-With-MPI/03-Accessing-System-Configurations-With-MPI.html#Using-Global-Snapshots-With-MPI) shows how you can access global snapshot data in an `if snapshot.communicator.rank == 0:`.

**Reply by sgoh446** [reply]
Thank you very much for your quick response!

So I guess the domain decomposition is performed when I create states from a snapshot, i.e., at
`simulation.create_state_from_snapshot(snapshot)
`

I modified my script accordingly and managed to finish test runs with multiple CPUs, which showed a nice speed-up (with AMD EPYC 7402):
1 core: 6.488936744076383 steps/second
2 cores: 12.326284516716738
3 cores: 23.845274689218574
4 cores: 46.83127312806327
8 cores: 82.27007826969572
16 cores: 144.22735423211734

However, tests with GPUs (A100) were not successful.
When the system size is large, or many GPUs are used, for instance, 
1 GPU: 23.81677374886617
2 GPU: 25.570778123892467
4 GPU: 38.43909963712529
but with 8 GPUs, I received the following error message:
`Caught signal 11 (Segmentation fault: address not mapped to object at address 0x14b054fed914)`

I have tested some more system sizes, and it seems like this error occurs when the system size is large or many GPUs are used. 
Here is the script I used:

```
import hoomd
import numpy

device = hoomd.device.GPU()
simulation = hoomd.Simulation(device=device, seed=1)

snapshot = hoomd.Snapshot()
L=128
density = 50
kT = 1

if snapshot.communicator.rank == 0:
    rng = numpy.random.default_rng(seed=42)
    snapshot.configuration.box = [L,L,L,0,0,0]
    snapshot.mpcd.types = ['A']
    snapshot.mpcd.N = int(density * L * L * L)
    snapshot.mpcd.position[:] = [0.5*L, 0.5*L, 0.5*L] * rng.uniform(low=-1,high=1,size=(snapshot.mpcd.N,3))

    velocity = rng.normal(0.0, numpy.sqrt(kT), (snapshot.mpcd.N, 3))
    velocity -= numpy.mean(velocity, axis=0)
    snapshot.mpcd.velocity[:] = velocity

simulation.create_state_from_snapshot(snapshot)

integrator = hoomd.mpcd.Integrator(dt=0.02)
integrator.collision_method = hoomd.mpcd.collide.StochasticRotationDynamics(
    period=1, angle=130, kT=kT
)

integrator.streaming_method = hoomd.mpcd.stream.Bulk(
    period=integrator.collision_method.period
)

integrator.mpcd_particle_sorter = hoomd.mpcd.tune.ParticleSorter(trigger=20)
simulation.operations.integrator = integrator

simulation.run(100)
device.notice(f'{simulation.tps}')
```

It seems to me that the system lost some particles.
Would you have any idea what is going on in the simulation?

Thank you!

**Reply by joaander** [reply]
@mphoward maintains the `mpcd` module, he can look into the segmentation fault.

I will point out that A100 GPUs are massive. Scaling to more than one GPU will require an extremely large number of particles. Try plotting performance vs system size on 1 GPU and you will see that very large systems are needed to even saturate a single GPU.

**Reply by mphoward** [reply]
I'll be happy to help!

Two quick comments first:

1. I agree with @joaander that you will probably need a pretty large system size to get decent scaling to multiple GPUs, especially for something like A100.
2. The performance for 1 GPU seems way too low compared to 1 CPU core. It could be that you are using a different system for the GPU benchmark or that the GPU is not saturated. However, you should also make sure that you run a warmup period before measuring TPS when you are doing performance benchmarks with GPUs. HOOMD uses GPU runtime autotuning that can make a huge difference in performance once completed, but performance can be poor while tuning is happening. See here for a method and example about how to do that:
https://hoomd-blue.readthedocs.io/en/v4.8.2/package-hoomd.html#hoomd.Operations.is_tuning_complete

> I have tested some more system sizes, and it seems like this error occurs when the system size is large or many GPUs are used. 
> It seems to me that the system lost some particles. Would you have any idea what is going on in the simulation?

Particles out-of-bounds should handled as exceptions with nice error messages, so a segfault usually means some unexpected error that we need to investigate. Could you give me a little more information here to help with that:
* Specific box size(s) and MPCD density that produce segfaults
* For each system configuration, what number(s) of GPUs trigger the segfault
* Whether this segfault happens only on GPUs or also if you run on an equivalent number of CPUs

**Reply by sgoh446** [reply]
Dear Michael, dear Joshua,

Thank you for your comments, and suggestions.

1. As for the system size:
I would guess 256x256X256 with a density of 10 should be large enough.
With a density of 20, I got:
```
Traceback (most recent call last):
  File "/p/project1/cdim/hoomd/test_hoomd_mpcd.py", line 27, in <module>
    simulation.create_state_from_snapshot(snapshot)
  File "/p/project1/cdim/hoomd/hoomd/simulation.py", line 292, in create_state_from_snapshot
    self._state = State(self, snapshot, domain_decomposition)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/p/project1/cdim/hoomd/hoomd/state.py", line 263, in __init__
    self._cpp_sys_def = _hoomd.SystemDefinition(
                        ^^^^^^^^^^^^^^^^^^^^^^^^
MemoryError: std::bad_alloc
srun: error: jwb0129: task 0: Exited with exit code 1
```

2. Therefore, I proceeded with 256^3, density 10 (after tuning as you suggested):
1 GPU: 19.656384772159406
2 GPUs: 33.255252771097865
3 GPUs: Segmentation fault
4 GPUs: Segmentation fault

3. The same happens also with CPUs.
1 CPU: 0.10005542870639476
2 CPUs: 0.1881823126051833
3 or 4 CPUs: Segmentation fault

I tried different seed values, or a different system size (200^3, density 20), but still, segfault.

Many thanks,
Segun

**Reply by mphoward** [reply]
Thanks, these timings make much more sense!

For this error:

> MemoryError: std::bad_alloc
> srun: error: jwb0129: task 0: Exited with exit code 1

I suspect that the system size might be too large to allocate the snapshot (or related memory for initialization) on the root process. All of the HOOMD snapshot data resides on the root rank and then gets distributed to the other processes. This isn't usually a problem for typical MD simulations, but for MPCD with big boxes and high densities, it can be. In 2.9.7, we had [make_random](https://hoomd-blue.readthedocs.io/en/v2.9.7/module-mpcd-init.html#hoomd.mpcd.init.make_random) to help circumvent this issue. That didn't make it into the rerelease, but I would be happy to work with you on a pull request to reimplement it if you have a use case.

For the segfaults, I think the issue is also related to system size & initialization under MPI, but in a different way, along the lines of this comment:

https://github.com/glotzerlab/hoomd-blue/issues/774#issuecomment-1979411890

Could you please try taking your L = 256 box and decrease the density from 10 in small-ish steps until you don't get a segfault anymore? I estimate that it should not be too much less than 10, and you can use CPUs instead of GPUs if that's faster turnaround. Thanks!

**Reply by sgoh446** [reply]
Thank you for the comments and suggestions.

As you had suggested, I tried some tests, decreasing the density.
This time I used L=128, increasing the density from 40 to 50, with 48 CPUs.
Segmentation faults occurred from 44.

Before deciding how to proceed, I would like to clarify a couple of things:

1. You mentioned that "make_random" is included in 2.9.7.
Would this mean that, if I install the 2.9.7 version, I can use "make_random" on my machine?

2. If I use "make_random", will the segfault issue be resolved simultaneously?

Thank you very much!

**Reply by mphoward** [reply]
Thank you, this is very helpful! You are obtaining a segfault when the number of particle has increased to the point where the memory for holding the positions is roughly 2 GB (3 doubles / particle), which is where I would have expected it to occur. HOOMD has convenience MPI routines that serialize data into byte streams and sends it. However, the maximum count of data that standard MPI can send is a signed int, which means about 2 GB of 1-byte objects. This error is probably never seen in MD because there are typically far fewer particles.

If you don't mind waiting a day or two, let me see if I can write a quick patch to fix this for MPCD by using the MPI library directly on the data instead of serializing. That should fix the segfault issue, at least.

I wouldn't recommend trying to get 2.9.7 to compile if you are using A100. That release is several years old, and I have heard that it is hard (if not impossible) to compile for GPUs with dependencies typically available on current HPC systems.

**Reply by joaander** [reply]
> If you don't mind waiting a day or two, let me see if I can write a quick patch to fix this for MPCD by using the MPI library directly on the data instead of serializing. That should fix the segfault issue, at least.

Could we fix the helper code to send multiple messages when the serialized object is more than 2GB? This could also fix other unexpected problems in the future where code unknowingly prepares more than 2GB of serialized data. I realize this may involve additional `MPI_Send` calls. Perhaps we could enable the mode only when the caller requests it to maintain performance of existing code paths. At a minimum, could we add an exception to the helper so that the user obtains a useful error message in place of a segmentation fault?

**Reply by mphoward** [reply]
> Could we fix the helper code to send multiple messages when the serialized object is more than 2GB? This could also fix other unexpected problems in the future where code unknowingly prepares more than 2GB of serialized data. I realize this may involve additional `MPI_Send` calls. Perhaps we could enable the mode only when the caller requests it to maintain performance of existing code paths.

I thought about trying to do this, but I started to get a headache planning out how scatter / gather would work. I think we might be better off using `MPI_Datatype`s because I believe the limit in the MPI library is on the _count_ of what you send and not its byte size. For example, we can create a datatype for `Scalar3` and then we can send up to an int's worth of them (24x more than what we do by serialization). Technically we still run into an issue in HOOMD because `N` is an unsigned int and the MPI library uses a signed int but this is a far less likely scenario... it could be handled by batching from the caller if necessary. Using datatypes should also reduce memory demands on the root process because space to serialize/deserialize is not needed.

> At a minimum, could we add an exception to the helper so that the user obtains a useful error message in place of a segmentation fault?

Yes, I can definitely do this easily. The size of the serialized data is known, and we can throw an exception if it will exceed `std::numeric_limits<int>::max()`.

Let me open an issue with some of these thoughts in one spot. I think as a first pass, I can try using the MPI datatype for scattering the MPCD snapshot and add the exception to the helper methods. Longer term, we could then write overloads of the helper methods for datatypes that can be sent without serialization, with fallback to the serialized version (+ exception). This might be best done by moving the helper methods to the `MPIConfiguration` class, which can hold the registered MPI datatypes.

**Reply by mphoward** [reply]
@sgoh446 Could you try compiling the branch for this draft pull request:

https://github.com/glotzerlab/hoomd-blue/pull/1897

and let me know if it fixes the issue you were having for the L = 128 test case on 48 CPUs? Thanks!

**Reply by sgoh446** [reply]
Yes, it does, I was able to run L=128, density 50, and also L=256 with density 10.

Here is the speedup for L=256, density 10:
1 GPU: 17.584475932281656
2: 17.103018105468752
4: 23.089839622014704
8: 31.751673968096394
16: 40.11789445643914
32: 37.171577496698696
64: 55.32130058164815
128: 68.28837909337616
256: 101.95002358868672

which seems not as good as the CPU case.
It seems that the system size is still small, but I couldn't go beyond that due to the memory issue at initialization.

I don't have a particular use case, but I am doing MPCD benchmarks for my publication.
Could we try "make_random"? 

Thanks a lot!

**Reply by mphoward** [reply]
Great to hear! We'll get the PR polished up and merged soon then.

Yes, it doesn't surprise me that the scaling to that many A100s is poor for this system size. If you are able to wait a week or so, I should be able to get the random initialization code path working again (the C++ code is still there, but it is missing a Python interface) so that you can try a larger system. I would also welcome a contribution from you if you would like to help implement it!

By the way, if you have performed any profiling, I would love to have that data to know what the bottleneck is, if you are willing to share it. I have not revisited the scaling behavior of this software since my 2018 publication and that test used K20x GPUs that are vastly different in terms of compute density than A100. At the time, I think the cell communication was the limiting factor, and I had some ideas about how to improve that... but they were not worth pursuing with the hardware at the time.

**Reply by sgoh446** [reply]
We are revising our manuscript and the deadline is approaching.
For now, the CPU scalability seems good enough for my comparison purposes, and it does not suffer too much from memory issues.

As for the profiling: I haven't done anything in this direction, sorry.

Anyway, thank you very much for your help!

**Reply by mphoward** [reply]
OK, no problem and you're welcome!
